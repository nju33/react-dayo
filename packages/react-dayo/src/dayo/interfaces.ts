import {BlockComponent} from '../entities/seed';

export interface DayoOptions {
  to: 'top' | 'bottom';
  position: 'left' | 'center' | 'right';
  maxLength: number;
}

export type DayoProps = DayoOptions;

// export interface DayoStruct<T> {
//   state: {
//     queue: T[];
//   };
//   render(): JSX.Element;
// }

export interface DayoImpl<Seed> {
  getOption<Key extends keyof DayoProps>(key: Key): DayoProps[Key];
  getQueue(): Seed[];
  setQueue(queue: Seed[]): void;
  onTransitionEnd(seed: Seed): () => void;
}

export type IDayo<Seed> = DayoImpl<Seed>;

export interface DayoState<Seed> {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  queue: Seed[];
}

export interface DayoOperator<Seed> {
  /**
   * Update target seed state for running re-rendering
   */
  rewriteQueueItem(seed: Seed): void;
  /**
   * Return index number in queue by passed `id`
   */
  getQueueIndexById(id: string): number;
  /**
   * To add a seed into queue state
   */
  addSeed(seed: Seed): void;
  /**
   * To skip a overflowed seeds in queue greater than `maxLength`
   */
  skipOverflowSeeds(opts: {maxLength: number}): void;
}

export interface DayoSelector {
  getQueueComponentProps(): QueueComponentProps;
  getBoxComponentProps(): BoxComponentProps;
}

export interface QueueComponentProps {
  to: DayoOptions['to'];
  position: DayoOptions['position'];
}

export interface BoxComponentThemeParams {
  transitionTimingFunction: string;
}

export interface BoxComponentProps {
  BlockComponent: BlockComponent;
  theme: BoxComponentThemeParams;
  additionalProps?: object;
  to: DayoOptions['to'];
  isEnter: boolean;
  isEntering: boolean;
  isEntered: boolean;
  isExit: boolean;
  isExiting: boolean;
  isExited: boolean;
  close(): void;
  onTransitionEnd(): void;
}
