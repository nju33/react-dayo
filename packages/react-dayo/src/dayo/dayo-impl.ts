// import {DispatcherImpl} from '../dispatcher';

export type DayoOptionTo = 'top' | 'bottom';

export interface DayoOptions {
  to: DayoOptionTo;
  maxLength: number;
}

export type DayoProps = DayoOptions;

export interface DayoStruct<T> {
  state: {
    queue: T[];
  };
  render(): JSX.Element;
}

export type IDayo<T> = DayoStruct<T>;
