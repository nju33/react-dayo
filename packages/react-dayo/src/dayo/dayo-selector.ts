import {IDayo} from './interfaces';
import {ISeed} from '../entities/seed';
import {DayoSelectorImpl} from './interfaces';

export class DayoSelector implements DayoSelectorImpl {
  public dayo: IDayo<unknown>;

  public constructor(dayo: IDayo<unknown>) {
    this.dayo = dayo;
  }

  public getQueueComponentProps() {
    return {
      to: this.dayo.getOption('to'),
      position: this.dayo.getOption('position'),
    };
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public getBoxComponentProps(seedOnCycle: ISeed) {
    if (seedOnCycle.BlockComponent === undefined) {
      throw new Error('BlockComponent is undefined');
    }

    if (seedOnCycle.values === undefined) {
      throw new Error('Unexpected error');
    }

    return {
      to: this.dayo.getOption('to'),
      BlockComponent: seedOnCycle.BlockComponent,
      additionalProps: seedOnCycle.values.props,
      theme: seedOnCycle.theme,
      close: seedOnCycle.close,
      isEnter: seedOnCycle.cycle.isEnter(),
      isEntering: seedOnCycle.cycle.isEntering(),
      isEntered: seedOnCycle.cycle.isEntered(),
      isExit: seedOnCycle.cycle.isExit(),
      isExiting: seedOnCycle.cycle.isExiting(),
      isExited: seedOnCycle.cycle.isExited(),
      onTransitionEnd: this.dayo.onTransitionEnd(seedOnCycle),
    };
  }
}

export default DayoSelector;
