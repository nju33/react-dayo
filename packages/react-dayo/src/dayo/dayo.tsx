import React from 'react';
import {DayoOptions, DayoProps} from './interfaces';
import {IDayo} from './interfaces';
import {ISeed} from '../entities/seed';
import Dispatcher, {Event as DispatcherEvent} from '../use-cases/dispatcher';
import Queue from '../components/queue';
import Box from '../components/box/box';

export const defaultOptions = {
  to: 'top' as DayoOptions['to'],
  position: 'center' as DayoOptions['position'],
  maxLength: 5,
};

/**
 * To create the `Dayo` component and also`dispatch` function to pass a seed into `Dayo`
 */
export const createDayo = <
  BlockComponentProps extends object = {},
  BCP extends BlockComponentProps = BlockComponentProps
>(
  userOptions: Partial<DayoOptions> = {},
): [React.ComponentClass<Partial<DayoProps>>, Dispatcher['dispatch']] => {
  const options: DayoOptions = {...defaultOptions, ...userOptions};
  const dispatcher = new Dispatcher();

  /**
   * Interface
   */
  class Dayo extends React.Component<Partial<DayoProps>>
    implements IDayo<ISeed<BCP>> {
    public static defaultProps = options;

    public dispatcher = dispatcher;

    public state = {
      queue: [] as ISeed<BCP>[],
    };

    /**
     * To just return `option`
     */
    private getOption<Key extends keyof DayoProps>(key: Key): DayoProps[Key] {
      return (
        ((this.props as unknown) as DayoProps)[key] ||
        ((options as unknown) as DayoProps)[key]
      );
    }

    /**
     * To refresh all items of queue for re-rendering
     */
    private rewriteQueueItem(seedOnCycle: ISeed<BCP>): void {
      const targetIndex = this.state.queue.findIndex(
        (item): boolean => {
          return item.id === seedOnCycle.id;
        },
      );

      if (targetIndex === -1) {
        return;
      }

      const {queue} = this.state;
      queue[targetIndex] = seedOnCycle;

      this.setState({queue});
    }

    /**
     * To be called when rendered in the dom
     */
    public componentDidMount(): void {
      dispatcher.on(DispatcherEvent.UpdateSeed, this.handleUpdateSeed);
      dispatcher.on(DispatcherEvent.DoneSeed, this.handleDoneSeed);
    }

    /**
     * To be called when the cycle step is updated
     *
     * if the step of current is 'enter' then addind it into the queue,
     * or else just re-rendering with new state
     *
     */
    private handleUpdateSeed = (seedOnCycle: ISeed<BCP>): void => {
      if (seedOnCycle.cycle.isEnter()) {
        this.addAlert(seedOnCycle);
        return;
      }

      return this.rewriteQueueItem(seedOnCycle);
    };

    /**
     * To be called when the cycle step is ended for removing a ended seed from the queue
     */
    private handleDoneSeed = (seed: ISeed<BCP>): void => {
      const {queue} = this.state;
      const nextQueue = queue.filter(
        (item): boolean => {
          return item.id !== seed.id;
        },
      );

      this.setState({queue: nextQueue});
    };

    /**
     * To add new seed into the queue;
     */
    private addAlert(seedOnCycle: ISeed<BCP>): void {
      this.setState({
        queue: [...this.state.queue, seedOnCycle],
      });

      this.closeOverflowSeeds();
    }

    /**
     * To close seeds overlowed
     */
    private closeOverflowSeeds(): void {
      const queueMaxLength = this.getOption('maxLength');
      const enteredItems = this.state.queue.filter(
        (item): boolean => item.cycle.isEntering() || item.cycle.isEntered(),
      );
      const newItemLength = 1;

      const overflowLength =
        enteredItems.length + newItemLength - queueMaxLength;

      if (overflowLength > 0) {
        enteredItems.slice(0, overflowLength).forEach(
          (seedOnCycle): void => {
            seedOnCycle.cycle.skip();
          },
        );
      }
    }

    /**
     * Advance to the next cycle step at the css animation ended
     */
    private onTransitionEnd(seedOnCycle: ISeed<BCP>): () => void {
      return (): void => {
        if (seedOnCycle.cycle.isEntering()) {
          seedOnCycle.cycle.proceed();
          return;
        }

        if (seedOnCycle.cycle.isExiting()) {
          seedOnCycle.cycle.proceed();
        }
      };
    }

    /**
     * To just render component
     */
    public render(): JSX.Element {
      return (
        <Queue to={this.getOption('to')} position={this.getOption('position')}>
          {this.state.queue.map(
            (seedOnCycle): JSX.Element => {
              if (seedOnCycle.values === undefined) {
                throw new Error('Unexpected error');
              }

              return (
                <Box
                  key={seedOnCycle.id}
                  to={this.getOption('to')}
                  BlockComponent={seedOnCycle.BlockComponent}
                  additionalProps={seedOnCycle.values.props}
                  theme={seedOnCycle.theme}
                  close={seedOnCycle.close}
                  isEnter={seedOnCycle.cycle.isEnter()}
                  isEntering={seedOnCycle.cycle.isEntering()}
                  isEntered={seedOnCycle.cycle.isEntered()}
                  isExit={seedOnCycle.cycle.isExit()}
                  isExiting={seedOnCycle.cycle.isExiting()}
                  isExited={seedOnCycle.cycle.isExited()}
                  onTransitionEnd={this.onTransitionEnd(seedOnCycle)}
                >
                  <div>{seedOnCycle.message}</div>
                </Box>
              );
            },
          )}
        </Queue>
      );
    }
  }

  return [Dayo, dispatcher.dispatch];
};
