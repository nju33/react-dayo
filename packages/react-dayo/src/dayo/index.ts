export * from './interfaces';
export * from './dayo';
