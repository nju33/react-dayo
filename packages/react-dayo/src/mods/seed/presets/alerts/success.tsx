import React from 'react';
import styled, {css} from 'styled-components';
import {BlockComponentProps} from '../../seed-impl';
import SeedFactory from '../../seed-factory';
import {NAE, SHIRONERI} from '../../constants/colors';
import CloseIcon from '../../../../components/close-icon';

export const mixin = css`
  display: flex;
  align-items: center;
  font-size: .8em;
  color: '${SHIRONERI}';
  background: ${NAE};
  border: 1px solid ${NAE};
  padding: 0.25em 0.5em;
  margin-top: 0.3em;
  box-sizing: border-box;

  .middle {
    margin: 0 0.5em;
  }

  .right {
    svg {
      height: 0.7em;
      opacity: 0.7;
      transition: 0.2s;
      transition-property: height, opacity;
    }
    svg:hover {
      height: 0.8em;
      opacity: 1;
      cursor: pointer;
    }
  }
`;

const Alert = styled.div`
  ${mixin}
`;

// eslint-disable-next-line react/prop-types
const SuccessAlert: React.FC<BlockComponentProps> = (props): JSX.Element => {
  return (
    <Alert>
      <div className="middle">{props.children}</div>
      <div className="right">
        <CloseIcon onClick={props.close} />
      </div>
    </Alert>
  );
};

export default SeedFactory.create(
  {key: 'success', timeout: 5000},
  SuccessAlert,
);
