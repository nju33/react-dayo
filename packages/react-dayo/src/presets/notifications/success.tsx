import React from 'react';
import styled from 'styled-components';
import {Seed, BlockComponentProps} from 'core-dayo';
import mixin from './mixin';
import CloseIcon from '../../components/close-icon';

const SUMI = '#1c1c1c';
const ROKUSYOH = '#24936e';

const Notification = styled.div`
  ${mixin}
`;

// eslint-disable-next-line react/prop-types
const Success: React.FC<BlockComponentProps> = (props): JSX.Element => {
  return (
    <Notification
      theme={{
        textColor: SUMI,
        backgroundColor: ROKUSYOH,
      }}
    >
      <div className="dayo-notification--middle">{props.children}</div>
      <div className="dato-notification--right">
        <CloseIcon onClick={props.close} />
      </div>
    </Notification>
  );
};

export default new Seed(Success, {timeout: 5000}).builder;
