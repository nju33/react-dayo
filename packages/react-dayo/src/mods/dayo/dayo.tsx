import React from 'react';
import {DayoOptions, DayoOptionTo} from './dayo-impl';
import DayoImpl from './dayo-impl';
import {SeedFactoryImpl, SeedImpl} from '../seed';
import Dispatcher, {Event as DispatcherEvent} from '../dispatcher';

import Queue from '../components/queue';
import Box from '../components/box';

const defaultOptions = {
  to: 'top' as DayoOptionTo,
};

export interface DayoProps {
  maxLength?: number;
}

export const createDayo = <BlockComponentAdditionalProps extends object = {}>(
  userOptions: Partial<DayoOptions> = {},
): [
  React.ComponentClass<DayoProps>,
  Dispatcher<BlockComponentAdditionalProps>['dispatch']
] => {
  const options: DayoOptions = {...defaultOptions, ...userOptions};

  const dispatcher = new Dispatcher<BlockComponentAdditionalProps>();

  class Dayo extends React.Component<DayoProps>
    implements
      DayoImpl<
        SeedFactoryImpl<BlockComponentAdditionalProps>,
        SeedImpl<BlockComponentAdditionalProps>
      > {
    public dispatcher = dispatcher;

    public state = {
      queue: [] as SeedImpl<BlockComponentAdditionalProps>[],
    };

    private rewriteQueueItem(
      seedOnCycle: SeedImpl<BlockComponentAdditionalProps>,
    ): void {
      const targetIndex = this.state.queue.findIndex(
        (item): boolean => {
          return item.id === seedOnCycle.id;
        },
      );

      if (targetIndex === -1) {
        return;
      }

      const {queue} = this.state;
      queue[targetIndex] = seedOnCycle;

      this.setState({queue});
    }

    private handleUpdateSeed = (
      seedOnCycle: SeedImpl<BlockComponentAdditionalProps>,
    ): void => {
      if (seedOnCycle.cycle.isEnter()) {
        this.addAlert(seedOnCycle);
        // this.closeOverflowSeeds();
        return;
      }

      return this.rewriteQueueItem(seedOnCycle);
    };

    private handleDoneSeed = (
      seed: SeedImpl<BlockComponentAdditionalProps>,
    ): void => {
      const {queue} = this.state;
      const nextQueue = queue.filter(
        (item): boolean => {
          return item.id !== seed.id;
        },
      );

      this.setState({queue: nextQueue});
    };

    public componentDidMount(): void {
      dispatcher.on(DispatcherEvent.UpdateSeed, this.handleUpdateSeed);
      dispatcher.on(DispatcherEvent.DoneSeed, this.handleDoneSeed);
    }

    private addAlert(
      seedOnCycle: SeedImpl<BlockComponentAdditionalProps>,
    ): void {
      this.setState({
        queue: [...this.state.queue, seedOnCycle],
      });

      this.closeOverflowSeeds();
    }

    private closeOverflowSeeds(): void {
      const queueMaxLength = this.props.maxLength || 10;
      const enteredItems = this.state.queue.filter(
        (item): boolean => item.cycle.isEntering() || item.cycle.isEntered(),
      );
      const newItemLength = 1;

      const overflowLength =
        enteredItems.length + newItemLength - queueMaxLength;

      if (overflowLength > 0) {
        enteredItems.slice(0, overflowLength).forEach(
          (seedOnCycle): void => {
            seedOnCycle.cycle.skip();
          },
        );
      }
    }

    private onTransitionEnd(
      seedOnCycle: SeedImpl<BlockComponentAdditionalProps>,
    ): () => void {
      return (): void => {
        if (seedOnCycle.cycle.isEntering()) {
          seedOnCycle.cycle.proceed();
          return;
        }

        if (seedOnCycle.cycle.isExiting()) {
          seedOnCycle.cycle.proceed();
        }
      };
    }

    public render(): JSX.Element {
      return (
        <Queue to={options.to}>
          {this.state.queue.map(
            (seedOnCycle): JSX.Element => {
              return (
                <Box
                  key={seedOnCycle.id}
                  to={options.to}
                  Block={seedOnCycle.Block}
                  additionalProps={seedOnCycle.values.props}
                  theme={seedOnCycle.theme}
                  close={seedOnCycle.close}
                  isEnter={seedOnCycle.cycle.isEnter()}
                  isEntering={seedOnCycle.cycle.isEntering()}
                  isEntered={seedOnCycle.cycle.isEntered()}
                  isExit={seedOnCycle.cycle.isExit()}
                  isExiting={seedOnCycle.cycle.isExiting()}
                  isExited={seedOnCycle.cycle.isExited()}
                  onTransitionEnd={this.onTransitionEnd(seedOnCycle)}
                >
                  <div>{seedOnCycle.message}</div>
                </Box>
              );
            },
          )}
        </Queue>
      );
    }
  }

  return [Dayo, dispatcher.dispatch];
};
