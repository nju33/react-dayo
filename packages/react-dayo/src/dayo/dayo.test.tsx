import React from 'react';
import {DayoOptions} from 'core-dayo';
import {render, cleanup, wait, waitForElement} from '@testing-library/react';
import success from '../presets/notifications/success';
import {createDayo} from './dayo';
import 'jest-dom/extend-expect';

describe('dayo', () => {
  afterEach(cleanup);

  /* eslint-disable @typescript-eslint/no-explicit-any */
  let Dayo: React.ComponentClass<Partial<DayoOptions>, any>;
  let dispatch: any;
  /* eslint-enable @typescript-eslint/no-explicit-any */

  beforeEach(() => {
    const [_Dayo, _dispatch] = createDayo();
    Dayo = _Dayo;
    dispatch = _dispatch;
  });

  test('default', async () => {
    jest.setTimeout(10000);

    const {container, queryByTestId, queryByText, queryAllByText} = render(
      <Dayo />,
    );
    const message = 'test';

    expect(queryByTestId('dayo--queue')).toBeTruthy();
    expect(queryByTestId('dayo--box')).toBeFalsy();
    expect(queryByText(message)).toBeFalsy();
    expect(container.firstChild).toMatchSnapshot();

    dispatch(
      success()
        .message(message)
        .timeout(1500),
    )();
    const dayoBoxNode = await waitForElement(() => queryByTestId('dayo--box'));
    expect(dayoBoxNode).toBeTruthy();

    /**
     * dispatch two more times
     */
    dispatch(
      success()
        .message(message)
        .timeout(1500),
    )();
    dispatch(
      success()
        .message(message)
        .timeout(1500),
    )();
    await wait(() => new Promise(r => setTimeout(r, 1000)));
    const messages = queryAllByText(message);
    expect(messages).toHaveLength(3);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('attributes', () => {
    const {container, queryByTestId} = render(
      <Dayo to="bottom" position="right" />,
    );

    const queue = queryByTestId('dayo--queue');
    expect(queue).toBeTruthy();
    expect(queue).toHaveAttribute('data-to', 'bottom');
    expect(queue).toHaveAttribute('data-position', 'right');
    expect(container.firstChild).toMatchSnapshot();
  });
});
