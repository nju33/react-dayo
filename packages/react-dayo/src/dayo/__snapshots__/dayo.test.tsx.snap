// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dayo attributes 1`] = `
<aside
  class="sc-bwzfXH kLkFEs"
  data-position="right"
  data-testid="dayo--queue"
  data-to="bottom"
/>
`;

exports[`dayo default 1`] = `
<aside
  class="sc-bwzfXH kLkFEs"
  data-position="center"
  data-testid="dayo--queue"
  data-to="top"
/>
`;

exports[`dayo default 2`] = `
<aside
  class="sc-bwzfXH kLkFEs"
  data-position="center"
  data-testid="dayo--queue"
  data-to="top"
>
  <div
    aria-live="assertive"
    class="sc-htpNat iPDAVw"
    data-is-delete="false"
    data-is-deleted="false"
    data-is-deleting="false"
    data-is-enter="false"
    data-is-entered="false"
    data-is-entering="true"
    data-testid="dayo--box"
    data-to="top"
    style="height: 0px;"
  >
    <div>
      <div
        class="sc-bdVaJa CLsaZ"
      >
        <div
          class="dayo-notification--middle"
        >
          <div>
            test
          </div>
        </div>
        <div
          class="dato-notification--right"
        >
          <svg
            fill="none"
            height="1em"
            viewBox="0 0 500 500"
            width="1em"
          >
            <rect
              fill="black"
              height="592.306"
              rx="18.5096"
              transform="rotate(-45 27 53.1765)"
              width="37.0191"
              x="27"
              y="53.1765"
            />
            <rect
              fill="black"
              height="592.306"
              rx="18.5096"
              transform="matrix(-0.707107 -0.707107 -0.707107 0.707107 472 53.1765)"
              width="37.0191"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-live="assertive"
    class="sc-htpNat iPDAVw"
    data-is-delete="false"
    data-is-deleted="false"
    data-is-deleting="false"
    data-is-enter="false"
    data-is-entered="false"
    data-is-entering="true"
    data-testid="dayo--box"
    data-to="top"
    style="height: 0px;"
  >
    <div>
      <div
        class="sc-bdVaJa CLsaZ"
      >
        <div
          class="dayo-notification--middle"
        >
          <div>
            test
          </div>
        </div>
        <div
          class="dato-notification--right"
        >
          <svg
            fill="none"
            height="1em"
            viewBox="0 0 500 500"
            width="1em"
          >
            <rect
              fill="black"
              height="592.306"
              rx="18.5096"
              transform="rotate(-45 27 53.1765)"
              width="37.0191"
              x="27"
              y="53.1765"
            />
            <rect
              fill="black"
              height="592.306"
              rx="18.5096"
              transform="matrix(-0.707107 -0.707107 -0.707107 0.707107 472 53.1765)"
              width="37.0191"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-live="assertive"
    class="sc-htpNat iPDAVw"
    data-is-delete="false"
    data-is-deleted="false"
    data-is-deleting="false"
    data-is-enter="false"
    data-is-entered="false"
    data-is-entering="true"
    data-testid="dayo--box"
    data-to="top"
    style="height: 0px;"
  >
    <div>
      <div
        class="sc-bdVaJa CLsaZ"
      >
        <div
          class="dayo-notification--middle"
        >
          <div>
            test
          </div>
        </div>
        <div
          class="dato-notification--right"
        >
          <svg
            fill="none"
            height="1em"
            viewBox="0 0 500 500"
            width="1em"
          >
            <rect
              fill="black"
              height="592.306"
              rx="18.5096"
              transform="rotate(-45 27 53.1765)"
              width="37.0191"
              x="27"
              y="53.1765"
            />
            <rect
              fill="black"
              height="592.306"
              rx="18.5096"
              transform="matrix(-0.707107 -0.707107 -0.707107 0.707107 472 53.1765)"
              width="37.0191"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</aside>
`;
